### PR TITLE
Http controller

### PR DESCRIPTION
**Pull request by jfahrer**
HTTP endpoints for building (RESTful) APIs

### CODE REVIEW

**Comment by ryanstout** [comment]
This looks great.  Still reviewing, but had one quick question.  Any ideas why the capybara specs run so much slower?  Is that selenium-webdriver 2.45.0?

**Comment by ryanstout** [comment]
Yea, 2.43.0 is faster.  Weird.

**Comment by jfahrer** [comment]
Yes, there seems to be an issue with 2.45.0: http://code.google.com/p/selenium/issues/detail?id=8551

**Comment by ryanstout** [comment]
@jfahrer ok, I had time to go through everything.  Looks great.  Nice work!

I've been thinking a bit about what volt components will look like in the future.  The long term plan is to support mobile app development (via either rubymotion or something like javascriptcore that gives a bridge to native api's).  In a perfect world, it would be great to be able to share code between server, browser, ios, and android.  I personally much prefer to do mobile app development with native components when possible, so I don't see sharing view code.  But I do think for many apps it would be possible to share controller code and definitely share models.  So I've been thinking more about the folder structure.  I think we do need a good way to split out code that is shared from code that isn't.  (for loading/compilation)  So I was thinking we could make it so inside of any of the current folders you could place the following (server, browser, ios, android, etc...)  I was thinking it might look something like this:

Example Structure:
- assets
  - browser
    - js
    - css
  - images
- controllers
  - .. shared controllers .. (the current controllers)
  - server
    - the http controllers
- models
  ....
- views
  - browser
  - ios
  - android

Basically, in the build processes, browser/ios/android/server would all restrict it to only that during the build processes.  This would let you easily reuse things like controllers, models, and images, but then separate things when you need the separation.

So I'm still thinking through some of the details, but I wanted to get your thoughts on using a "server" folder inside of controllers for the HttpControllers?  Then we could raise an exception or something if a HttpController gets parsed in Opal for example. 

Please let me know your thoughts on things?  Thanks!

**Comment by jfahrer** [comment]
Sweet. Let me know if I can help you with something esle / enhance the controllers somehow.

Putting the current controllers + http controllers inside of "controllers" and use a subdirectoryis a good idea I think. And "server" better describes the usage of the controllers. Raising exceptions when things get transpiled by opal when the should not is also a good thing.

And some random thoughts (no need to answer them all ;) :
I'd like to see something like a lib directory under apps for every component. This would allow me to put custom classes to the component they belong. Maybe use a shared and server directory here too?

I'd call it "web" instead of "browser".

Maybe we could use the main view folder for web/browser stuff and have sub directories for android and ios? Though you couldn't have controllers/views named that way then.

Might it be useful to have the abillity to extend certain controllers somehow for ios / android if more logic is required for something? Some applications want to access data different on the mobile side (number of entries for paginations for example).

If I would acces a model from the mobile device  - it would load the model with all of it's fields. Wouldn't it? This might not be desirable when it comes to saving bandwith. Maybe handled by the permissions dsl?

Btw. - making volt a framework for building mobile and web applications inside one project is great thing! Especially if it will be possible to reuse controllers and models - and let volt take care of the data transportation! Usally i'd prefer an API first approach when i would have to build an application that has to expose all it's capabilities for different front ends and then use an rich client side framework for the web stuff. Getting the "API for free" is a neat thing. It will also reduce the amount of duplicated code for validations an such.

I'm totaly with you when it comes using native components for mobile apps.

**Comment by ryanstout** [comment]
@jfahrer I'll need to think about web over browser.  Its shorter, so I like that.  In the mean time, would you be able to move the folder for the http controllers to controllers/server?  Let me know if I can help.  

Also, we should add ```lib``` to components.  I just hadn't gotten around to doing it.  Basically, we just need to add it to the generators, then make sure it gets added to the load path.  If you wanted to try and tackle that one let me know.  Opal has its own load paths that you use Opal.append_path to add to.  I can take it if you have other things to work on.

Also, I'm planning on supporting something like .select to limit what fields get loaded.  I need to think it through a bit more first though.  I'm working through some details in my head to allow one component to receive a query, then add onto that query so that all of the data an app needs for a page can get loaded in a single trip.

Yea, I think once we get the rest stuff into place, we could make a gem that gives you an auto-generated rest api.  Since the validation/permission logic is in the models, then its just a matter of hooking up the authentication.

Thanks!

**Comment by ryanstout** [comment]
Also, if you have time, could you do a PR for the docs repo on the http stuff?  Lastly, does I noticed you did "plain" for when you want to output a string.  I think rails does "text", which I kind of like better personally.  (Though "string" might make the most since, hehe)  Would you mind changing it to text?

**Comment by jfahrer** [comment]
OK, I moved the http_controllers to controllers/server
```render plain:``` has been changed to ```render text:```
and amended the commit.

I'm going to write the docs tomorrow. I think I'll also have to time get started with the ```lib``` stuff
